File: Java/app/src/main/java/com/github/shin1x1/jsonparser/parser/Parser.java
package com.github.shin1x1.jsonparser.parser;

import com.github.shin1x1.jsonparser.lexer.Lexer;
import com.github.shin1x1.jsonparser.lexer.Token;
import com.github.shin1x1.jsonparser.lexer.exception.UnexpectedEotException;
import com.github.shin1x1.jsonparser.parser.exception.NoKeyException;
import com.github.shin1x1.jsonparser.parser.exception.UnexpectedTokenException;

import javax.annotation.Nonnull;
import java.io.IOException;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.Optional;

public final class Parser {
    @Nonnull
    private final Lexer lexer;

    public Parser(@Nonnull Lexer lexer) {
        this.lexer = lexer;
    }

    public JsonValue parse() throws IOException {
        var token = lexer.getNextToken();
        if (token.isEmpty()) {
            return new JsonValue.Null();
        }

        return parseValue(token.get());
    }

    private JsonValue parseValue(Token token) throws IOException {
        if (token instanceof Token.True) {
            return new JsonValue.True();
        } else if (token instanceof Token.False) {
            return new JsonValue.False();
        } else if (token instanceof Token.Null) {
            return new JsonValue.Null();
        } else if (token instanceof Token.String) {
            return new JsonValue.String(((Token.String) token).value());
        } else if (token instanceof Token.Number) {
            return new JsonValue.Number(((Token.Number) token).value());
        } else if (token instanceof Token.LeftBracket) {
            return parseArray();
        } else if (token instanceof Token.LeftBrace) {
            return parseObject();
        }

        throw new UnexpectedTokenException(token);
    }

    private JsonValue.Array parseArray() throws IOException {
        enum State {
            Default, Value, Comma
        }

        var list = new ArrayList<JsonValue>();
        var state = State.Default;

        while (true) {
            var token = lexer.getNextToken().orElseThrow(UnexpectedEotException::new);

            switch (state) {
                case Default -> {
                    if (token instanceof Token.RightBracket) {
                        return new JsonValue.Array(list);
                    }
                    list.add(parseValue(token));
                    state = State.Value;
                }
                case Value -> {
                    if (token instanceof Token.RightBracket) {
                        return new JsonValue.Array(list);
                    } else if (token instanceof Token.Comma) {
                        state = State.Comma;
                    } else {
                        throw new UnexpectedTokenException(token);
                    }
                }
                case Comma -> {
                    list.add(parseValue(token));
                    state = State.Value;
                }
            }
        }
    }

    private JsonValue.Object parseObject() throws IOException {
        enum State {
            Default, Key, Colon, Value, Comma
        }

        var map = new HashMap<String, JsonValue>();
        Optional<String> key = Optional.empty();
        var state = State.Default;

        while (true) {
            var token = lexer.getNextToken().orElseThrow(UnexpectedEotException::new);

            switch (state) {
                case Default -> {
                    if (token instanceof Token.RightBrace) {
                        return new JsonValue.Object(map);
                    }

                    if (token instanceof Token.String) {
                        key = Optional.of(((Token.String) token).value());
                        state = State.Key;
                    } else {
                        throw new UnexpectedTokenException(token);
                    }
                }
                case Key -> {
                    if (token instanceof Token.Colon) {
                        state = State.Colon;
                    } else {
                        throw new UnexpectedTokenException(token);
                    }
                }
                case Colon -> {
                    var value = parseValue(token);
                    var k = key.orElseThrow(NoKeyException::new);

                    map.put(k, value);
                    key = Optional.empty();
                    state = State.Value;
                }
                case Value -> {
                    if (token instanceof Token.RightBrace) {
                        return new JsonValue.Object(map);
                    }

                    if (token instanceof Token.Comma) {
                        state = State.Comma;
                    } else {
                        throw new UnexpectedTokenException(token);
                    }
                }
                case Comma -> {
                    if (token instanceof Token.String) {
                        key = Optional.of(((Token.String) token).value());
                        state = State.Key;
                    } else {
                        throw new UnexpectedTokenException(token);
                    }
                }
            }
        }
    }
}

File: Scala3/src/main/scala/parser/Parser.scala
package parser

import lexer.{Lexer, Scanner, Token}

import scala.annotation.tailrec
import scala.util.{Failure, Success, Try}

enum JsonValue:
  case Number(value: Double)
  case String(value: scala.Predef.String)
  case Null
  case True
  case False
  case Array(value: List[JsonValue])
  case Object(value: Map[scala.Predef.String, JsonValue])

type TryJsonValue = Try[JsonValue]

object Parser:
  def parse(scanner: Scanner): TryJsonValue =
    if scanner.isEot() then return Success(JsonValue.Null)

    Lexer.getNextToken(scanner) match
      case Success(t) => parseValue(t)(using scanner)
      case Failure(e) => Failure(e)

  private def parseValue(token: Token)(using scanner: Scanner): TryJsonValue = token match
    case Token.Number(v) => Success(JsonValue.Number(v))
    case Token.String(v) => Success(JsonValue.String(v))
    case Token.Null => Success(JsonValue.Null)
    case Token.True => Success(JsonValue.True)
    case Token.False => Success(JsonValue.False)
    case Token.LeftBracket => parseArray()
    case Token.LeftBrace => parseObject()
    case x => Failure(Exception(s"Unexpected Token: $x"))

  private def parseArray()(using scanner: Scanner): TryJsonValue =
    enum State:
      case Default, Value, Comma

    import State.*

    @tailrec
    def parse(state: State, list: List[JsonValue]): Option[List[JsonValue]] =
      val token = Lexer.getNextToken(scanner).getOrElse(return None)

      state match
        case Default => token match
          case Token.RightBracket => Some(list)
          case _ => parseValue(token) match
            case Success(v) => parse(Value, list :+ v)
            case Failure(_) => None
        case Value => token match
          case Token.RightBracket => Some(list)
          case Token.Comma => parse(Comma, list)
          case _ => None
        case Comma => parseValue(token) match
          case Success(v) => parse(Value, list :+ v)
          case Failure(_) => None

    parse(Default, List()) match
      case Some(l) => Success(JsonValue.Array(l))
      case None => Failure(Exception("Invalid array"))

  private def parseObject()(using scanner: Scanner): TryJsonValue =
    enum State:
      case Default, Value, Comma, Colon, Key

    import State.*

    @tailrec
    def parse(state: State, map: Map[String, JsonValue], key: Option[String]): Option[Map[String, JsonValue]] =
      val token = Lexer.getNextToken(scanner).getOrElse(return None)

      state match
        case Default => token match
          case Token.RightBrace => Some(map)
          case Token.String(s) => parse(Key, map, Some(s))
          case _ => None
        case Key => token match
          case Token.Colon => parse(Colon, map, key)
          case _ => None
        case Colon => {
          val keyString = key.getOrElse(return None)
          parseValue(token) match
            case Success(v) => parse(Value, map + (keyString -> v), None)
            case Failure(_) => None
        }
        case Value => token match
          case Token.RightBrace => Some(map)
          case Token.Comma => parse(Comma, map, None)
          case _ => None
        case Comma => token match
          case Token.RightBrace => Some(map)
          case Token.String(s) => parse(Key, map, Some(s))
          case _ => None

    parse(Default, Map(), None) match
      case Some(l) => Success(JsonValue.Object(l))
      case None => Failure(Exception("Invalid object"))
